jetbrains-core/src/software/aws/toolkits/jetbrains/core/credentials/ConnectionSettingsMenuBuilder.kt

```diff
@@ -80,7 +87,22 @@ class ConnectionSettingsMenuBuilder private constructor() {
                             override fun update(e: AnActionEvent) {
                                 super.update(e)
                                 if (value.lazyIsUnauthedBearerConnection()) {
-                                    e.presentation.icon = AllIcons.General.Warning
+                                    ApplicationManager.getApplication().executeOnPooledThread {
+                                        val provider = (value.getConnectionSettings().tokenProvider.delegate as? BearerTokenProvider)
+                                        if (provider != null && provider.state() == BearerTokenAuthState.NEEDS_REFRESH) {
+                                            try {
+                                                provider.resolveToken()
+                                                BearerTokenProviderListener.notifyCredUpdate(provider.id)
+                                            } catch (error: Exception) {
+                                                LOG.info {"Failed to refresh access token $error"}
+                                            }
+                                        }
+                                        // show warning icon if connection is in not authorized state.
+                                        if (provider?.state() != BearerTokenAuthState.AUTHORIZED) {
+                                            e.presentation.icon = AllIcons.General.Warning
+                                        }
```

Comment:
we should move L102 out of `executeOnPoolThread`

Reply:
and call a `get()` like this `ApplicationManager.getApplication().executeOnPooledThread {}.get()`

Reply:
since `executeOnPooledThread` returns a `Future`(similar to async/await) so we just await here by calling `get()`

Reply:
@Will-ShaoHua will `get()` block the `update` execution?

Reply:
this can run once a second so this is probably not the right place for this

Reply:
maybe we should just did what you previously did @leigaol 
```suggestion
                                        if (provider?.state() == BearerTokenAuthState.NOT_AUTHENTICATED) {
                                            e.presentation.icon = AllIcons.General.Warning
                                        }
```

Reply:
and add a `tryRefresh` somewhere else

Reply:
when do we update this connectionMenu?

Reply:
I think let's stick with the original change here in this PR
```
if (provider?.state() != BearerTokenAuthState.AUTHORIZED) {
                                            e.presentation.icon = AllIcons.General.Warning
                                        }
```

Reply:
This is the place where the "More" action icon in the aws toolkit panel `...` updates its icon.
```
class MoreConnectionActionsAction : DumbAwareAction(AllIcons.Actions.MoreHorizontal) {
    override fun update(e: AnActionEvent) {
        e.presentation.icon = if (lazyGetUnauthedBearerConnections().isEmpty()) AllIcons.Actions.MoreHorizontal else AllIcons.General.Warning
    }
    ```

the place here where your change has made is where the "Balloon Warning" icon can appear
```
override fun update(e: AnActionEvent) {
                                super.update(e)
                                if (value.lazyIsUnauthedBearerConnection()) {
                                    e.presentation.icon = AllIcons.General.Warning
```

So, it seems like these two places both use `lazyIsUnauthedBearerConnection()` to check if it needs to have a balloon warning icon. I would make one of the two changes below:
1. in each of the above logic, do manual refresh once(non-blocking for EDT) and update the icon to warning only after the refresh fails. 
2. like what you have initially changed, relax the condition such that if it's accessToken expired it would still show ok icon instead of the warning icon. And let the applications refresh token on usage(CW etc.)

Reply:
besides the change above, can we spread one more `CodeWhispererUtil.promptReAuth` at `CodeWhispererService`?

Reply:
IIRC we used to have this when user tries to trigger getRecommendation

Reply:
right now we only have this at projectStartup

Reply:
we still have this on line 91 of the file CodeWhispererService.kt if that is what you are talking about.

Reply:
oh ok found it

Reply:
OK I just reverted it back to 

> like what you have initially changed, relax the condition such that if it's accessToken expired it would still show ok icon instead of the warning icon. And let the applications refresh token on usage(CW etc.)


Doing an API call in a UX update function is expensive and not ideal.

Reply:
can you also change the relax CW checking condition then so it matches the changes here?

Reply:
Do you mean this code? 
`
    fun checkActiveCodeWhispererConnectionType(project: Project) = when {
        actionState.token != null -> CodeWhispererLoginType.Accountless
        isAccessTokenExpired(project) || isRefreshTokenExpired(project) -> CodeWhispererLoginType.Expired`

Reply:
yes